README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cranlogs <a href='https://r-hub.github.io/cranlogs'><img src='man/figures/logo.png' align="right" height="138.5" /></a>

> Download logs from the RStudio CRAN mirror

<!-- badges: start -->

[![R build
status](https://github.com/r-hub/cranlogs/workflows/rworkflows/badge.svg)](https://github.com/r-hub/cranlogs/actions)
[![Windows Build
status](https://ci.appveyor.com/api/projects/status/github/metacran/cranlogs?svg=true)](https://ci.appveyor.com/project/gaborcsardi/cranlogs)
[![CRAN
version](http://www.r-pkg.org/badges/version/cranlogs)](http://www.r-pkg.org/pkg/cranlogs)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/cranlogs)](http://www.r-pkg.org/pkg/cranlogs)
[![cran
checks](https://cranchecks.info/badges/worst/cranlogs)](https://cranchecks.info/pkgs/cranlogs)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Gitter
chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/r-hub/community)
<!-- badges: end -->

[RStudio](http://www.rstudio.com) publishes the download logs from their
CRAN package mirror daily at <http://cran-logs.rstudio.com>.

This R package queries a [web API maintained by
R-hub](https://github.com/r-hub/cranlogs.app#the-api-of-the-cran-downloads-database)
serving a database, also maintained by R-hub, that contains the daily
download numbers for each package.

The RStudio CRAN mirror is not the only CRAN mirror, but it’s a popular
one: it’s the default choice for RStudio users. The actual number of
downloads over all CRAN mirrors is unknown.

## Installation

You can install `cranlogs` from CRAN:

``` r
install.packages("cranlogs")
```

Or get the development version from Github:

``` r
# install.packages("remotes")
remotes::install_github("r-hub/cranlogs")
```

## Usage

It has a very simple API. By default it shows the total number of
package downloads from the last day for which data is available.

``` r
library(cranlogs)
cran_downloads()
```

    #>         date   count
    #> 1 2022-12-18 2996182

The last week (6 days prior to the last day for which data is
available), or the last month (29 days prior to the last day for which
data is available) can be also easily queried:

``` r
cran_downloads(when = "last-week")
```

    #>         date   count
    #> 1 2022-12-12 6056710
    #> 2 2022-12-13 6337076
    #> 3 2022-12-14 6073282
    #> 4 2022-12-15 5799764
    #> 5 2022-12-16 5442572
    #> 6 2022-12-17 3125759
    #> 7 2022-12-18 2996182

The `package` argument queries a specific package:

``` r
cran_downloads(when = "last-week", package = c("magrittr", "dplyr"))
```

    #>          date count  package
    #> 1  2022-12-12 36410 magrittr
    #> 2  2022-12-13 40604 magrittr
    #> 3  2022-12-14 36484 magrittr
    #> 4  2022-12-15 33027 magrittr
    #> 5  2022-12-16 29838 magrittr
    #> 6  2022-12-17 19336 magrittr
    #> 7  2022-12-18 18437 magrittr
    #> 8  2022-12-12 67459    dplyr
    #> 9  2022-12-13 65243    dplyr
    #> 10 2022-12-14 83682    dplyr
    #> 11 2022-12-15 64705    dplyr
    #> 12 2022-12-16 67751    dplyr
    #> 13 2022-12-17 33317    dplyr
    #> 14 2022-12-18 31816    dplyr

Alternative intervals can also be specified:

``` r
cran_downloads(package = "plyr", from = "2014-01-01", to = "2014-02-01")
```

    #>          date count package
    #> 1  2014-01-01   292    plyr
    #> 2  2014-01-02   604    plyr
    #> 3  2014-01-03   637    plyr
    #> 4  2014-01-04   403    plyr
    #> 5  2014-01-05   430    plyr
    #> 6  2014-01-06   870    plyr
    #> 7  2014-01-07  1044    plyr
    #> 8  2014-01-08  1158    plyr
    #> 9  2014-01-09  1178    plyr
    #> 10 2014-01-10  1092    plyr
    #> 11 2014-01-11   489    plyr
    #> 12 2014-01-12   560    plyr
    #> 13 2014-01-13  1274    plyr
    #> 14 2014-01-14  1323    plyr
    #> 15 2014-01-15  1434    plyr
    #> 16 2014-01-16  1379    plyr
    #> 17 2014-01-17  1094    plyr
    #> 18 2014-01-18   541    plyr
    #> 19 2014-01-19   564    plyr
    #> 20 2014-01-20  1206    plyr
    #> 21 2014-01-21  1474    plyr
    #> 22 2014-01-22  1572    plyr
    #> 23 2014-01-23  1556    plyr
    #> 24 2014-01-24  1541    plyr
    #> 25 2014-01-25   766    plyr
    #> 26 2014-01-26   823    plyr
    #> 27 2014-01-27  1536    plyr
    #> 28 2014-01-28  1551    plyr
    #> 29 2014-01-29  1558    plyr
    #> 30 2014-01-30  1535    plyr
    #> 31 2014-01-31  1181    plyr
    #> 32 2014-02-01   662    plyr

For downloads of R itself, give `"R"` instead of a package:

``` r
cran_downloads("R")
```

    #>          date      version  os count
    #> 1  2022-12-18       2.10.1 osx     1
    #> 2  2022-12-18       2.11.0 osx     1
    #> 3  2022-12-18       2.11.1 osx     1
    #> 4  2022-12-18       2.12.1 osx     1
    #> 5  2022-12-18       2.13.1 osx     3
    #> 6  2022-12-18       2.14.1 osx     1
    #> 7  2022-12-18       2.15.1 osx     1
    #> 8  2022-12-18        3.2.1 osx     4
    #> 9  2022-12-18        3.2.4 osx     2
    #> 10 2022-12-18        3.3.1 osx     1
    #> 11 2022-12-18        3.3.2 osx     1
    #> 12 2022-12-18        3.3.3 osx    32
    #> 13 2022-12-18        3.4.4 osx     4
    #> 14 2022-12-18        3.5.3 osx     4
    #> 15 2022-12-18        3.6.1 osx     1
    #> 16 2022-12-18        3.6.2 osx     1
    #> 17 2022-12-18        3.6.3 osx    50
    #> 18 2022-12-18     3.6.3.nn osx    33
    #> 19 2022-12-18        4.0.2 osx     6
    #> 20 2022-12-18        4.0.4 osx     2
    #> 21 2022-12-18        4.0.5 osx     6
    #> 22 2022-12-18        4.1.0 osx     4
    #> 23 2022-12-18        4.1.1 osx    64
    #> 24 2022-12-18        4.1.2 osx  1179
    #> 25 2022-12-18        4.1.3 osx   715
    #> 26 2022-12-18        4.2.0 osx     3
    #> 27 2022-12-18        4.2.1 osx    19
    #> 28 2022-12-18        4.2.2 osx  4898
    #> 29 2022-12-18       latest osx     1
    #> 30 2022-12-18      release osx     9
    #> 31 2022-12-18       2.12.2 src     1
    #> 32 2022-12-18        2.3.0 src     1
    #> 33 2022-12-18        2.7.0 src     1
    #> 34 2022-12-18        3.1.3 src     1
    #> 35 2022-12-18        3.2.4 src     1
    #> 36 2022-12-18        3.2.5 src     1
    #> 37 2022-12-18        3.3.0 src     1
    #> 38 2022-12-18        3.4.0 src     1
    #> 39 2022-12-18        3.4.1 src     1
    #> 40 2022-12-18        3.4.2 src     1
    #> 41 2022-12-18        3.5.2 src     4
    #> 42 2022-12-18        3.5.3 src     3
    #> 43 2022-12-18        3.6.3 src     4
    #> 44 2022-12-18        4.0.0 src     2
    #> 45 2022-12-18        4.0.1 src     1
    #> 46 2022-12-18        4.0.2 src     1
    #> 47 2022-12-18        4.1.1 src     5
    #> 48 2022-12-18        4.1.2 src     4
    #> 49 2022-12-18        4.1.3 src     6
    #> 50 2022-12-18        4.2.0 src    15
    #> 51 2022-12-18        4.2.1 src     8
    #> 52 2022-12-18        4.2.2 src   330
    #> 53 2022-12-18       latest src     1
    #> 54 2022-12-18        3.4.1 win     1
    #> 55 2022-12-18        3.4.2 win     1
    #> 56 2022-12-18        3.4.3 win     2
    #> 57 2022-12-18        3.4.4 win     2
    #> 58 2022-12-18        3.5.3 win     2
    #> 59 2022-12-18        3.6.0 win     3
    #> 60 2022-12-18        3.6.1 win     4
    #> 61 2022-12-18        3.6.2 win     1
    #> 62 2022-12-18        3.6.3 win   122
    #> 63 2022-12-18        4.0.0 win     3
    #> 64 2022-12-18        4.0.1 win     1
    #> 65 2022-12-18        4.0.2 win     3
    #> 66 2022-12-18        4.0.3 win    35
    #> 67 2022-12-18        4.0.4 win     3
    #> 68 2022-12-18        4.0.5 win     9
    #> 69 2022-12-18        4.1.0 win     9
    #> 70 2022-12-18        4.1.1 win     5
    #> 71 2022-12-18        4.1.2 win     4
    #> 72 2022-12-18        4.1.3 win   831
    #> 73 2022-12-18        4.2.0 win    27
    #> 74 2022-12-18        4.2.1 win    16
    #> 75 2022-12-18        4.2.2 win 47464
    #> 76 2022-12-18 4.2.2patched win    17
    #> 77 2022-12-18        devel win   798
    #> 78 2022-12-18  patched.exe win     1

## Top downloaded packages

Last day for which data is available:

``` r
cran_top_downloads()
```

    #>    rank     package count       from         to
    #> 1     1     ggplot2 61663 2022-12-18 2022-12-18
    #> 2     2         cli 51625 2022-12-18 2022-12-18
    #> 3     3        ragg 48820 2022-12-18 2022-12-18
    #> 4     4       rlang 48662 2022-12-18 2022-12-18
    #> 5     5 textshaping 48466 2022-12-18 2022-12-18
    #> 6     6   lifecycle 43302 2022-12-18 2022-12-18
    #> 7     7       vctrs 40830 2022-12-18 2022-12-18
    #> 8     8    devtools 38367 2022-12-18 2022-12-18
    #> 9     9          sf 36230 2022-12-18 2022-12-18
    #> 10   10       rgeos 34583 2022-12-18 2022-12-18

Last week (6 days prior to the last day for which data is available):

``` r
cran_top_downloads("last-week")
```

    #>    rank     package  count       from         to
    #> 1     1         cli 604923 2022-12-12 2022-12-18
    #> 2     2       rlang 597346 2022-12-12 2022-12-18
    #> 3     3     ggplot2 559945 2022-12-12 2022-12-18
    #> 4     4   lifecycle 524145 2022-12-12 2022-12-18
    #> 5     5       vctrs 502146 2022-12-12 2022-12-18
    #> 6     6       dplyr 413973 2022-12-12 2022-12-18
    #> 7     7        ragg 371705 2022-12-12 2022-12-18
    #> 8     8 textshaping 365632 2022-12-12 2022-12-18
    #> 9     9      pillar 312270 2022-12-12 2022-12-18
    #> 10   10    devtools 305118 2022-12-12 2022-12-18

## Raw JSON API

See [the GitHub repo of the API of the CRAN downloads
database](https://github.com/r-hub/cranlogs.app).


NEWS.md

# cranlogs 2.1.2

* Fix `cran_downloads` and prevent it from failing with many packages:
  https://github.com/r-hub/cranlogs/issues/56
* Speed up `cran_downloads` with parallelisation.
* Add new unit test for `cran_downloads` with lots of packages.
* Remove `dontrun{}` from `cran_downloads` and `cran_top_downloads` examples.
  Unclear why this was here?
* New support function: `message_parallel`, `split_batches`
* Replace travis (failing)  and pkgdown with `rworkflows`.
* Update Authors in *DESCRIPTION* to reflect contribution, make fields explicit.
* Add *node_modules$* to *.Rbuildignore*.

# cranlogs 2.1.1

* New function `cranlogs_badge()` to construct the Markdown string of a
  badge for the README.

* Update the definition of the argument values "last-day", "last-week" and
  "last-month" to make them consistent with the database: "last-day" is
  the last day for which data is available, "last-week" 6 days prior to
  that last day with data, "last-month" 29 days prior to that last day
  with data.

* Drop row names after reordering which avoids having meaningless row
  names (@patperry, #23)

* Fix R download counts if a day is missing data (e.g. 2015-08-23).

# cranlogs 2.1.0

* Query downloads of R: `cran_downloads("R")`.

# cranlogs 2.0.0

* First release on CRAN.
* Queries return data frames.

# cranlogs 1.0.0

* First public release.


tests/testthat.R
library(testthat)
library(cranlogs)

test_check("cranlogs")


tests/testthat/test-badge.R
context("test-badge")

test_that("cranlogs_badge works", {
  expect_equal(cranlogs_badge("praise"),
               "[![CRAN RStudio mirror downloads](https://cranlogs.r-pkg.org/badges/last-month/praise?color=blue)](https://r-pkg.org/pkg/praise)")
  expect_equal(cranlogs_badge("praise", summary = "last-week", color = "orange")
, "[![CRAN RStudio mirror downloads](https://cranlogs.r-pkg.org/badges/last-week/praise?color=orange)](https://r-pkg.org/pkg/praise)")
})

test_that("parameter checks", {
  expect_error(cranlogs_badge(), "no default")
  expect_error(cranlogs_badge("praise", summary = "all"),
               'should be one')
})


tests/testthat/test-check_date.R
context("test-check_date")

test_that("check_date throws an error if the date is poorly formatted", {
  expect_error(check_date("lala"), "formatted")
  expect_error(check_date("2016-31-12"), "formatted")
  expect_error(cran_downloads(from = "2019-01-01", to = "2019-31-12"), 
               "formatted")
  expect_error(cran_downloads(from = "2019-99-01", to = "2019-01-12"), 
               "formatted")
})

test_that("check_date throws no error if the date is ok", {
  from <- as.Date("2019-02-23", format = "%Y-%m-%d")
  expect_silent(check_date(from))
  
  expect_silent(check_date("2019-02-23"))
})


R/badge.R
#' Create Markdown code for a cranlogs badge
#'
#' @param package_name name of the package 
#' @param summary One of "last-day", "last-week", "last-month", "grand-total".
#' @param color Either a hex color code or one of "brightgreen", "green", "yellowgreen",
#'  "yellow", "orange", "red", "lightgrey", "blue".
#'
#' @return The string to be copy-pasted to the README source.
#' @export
#'
#' @examples
#' cranlogs_badge("praise")
#' cranlogs_badge("praise", summary = "grand-total")
#' cranlogs_badge("praise", summary = "last-week", color = "orange")
#' cranlogs_badge("praise", color = "ff8c69")
cranlogs_badge <- function(package_name, 
                           summary = c("last-month", "last-day", 
                                       "last-week", "grand-total"),
                         color = "blue"){
  
  summary <- match.arg(summary)
  
  pkg_url <- paste0("https://r-pkg.org/pkg/", package_name)
  
  badge_url <- paste0("https://cranlogs.r-pkg.org/badges/",
               summary, "/",
               package_name, 
               "?color=", color)
  
  paste0("[![CRAN RStudio mirror downloads](",
         badge_url, ")](", pkg_url, ")")
  
}

project_name = function(base_path = usethis::proj_get()) {
  # taken from usethis
  proj_crit = function ()  {
    rprojroot::has_file(".here") | 
      rprojroot::is_rstudio_project | 
      rprojroot::is_r_package | 
      rprojroot::is_git_root | 
      rprojroot::is_remake_project | 
      rprojroot::is_projectile_project
  }
  
  proj_find = function(path = ".") {
    tryCatch(rprojroot::find_root(proj_crit(), path = path), 
             error = function(e) NULL)
  }
  possibly_in_proj = function (path = ".") !is.null(proj_find(path))
  if (!possibly_in_proj(base_path)) {
    return(fs::path_file(base_path))
  }
  
  is_package = function (base_path = usethis::proj_get()) 
  {
    res <- tryCatch(rprojroot::find_package_root_file(path = base_path), 
                    error = function(e) NULL)
    !is.null(res)
  }
  if (is_package(base_path)) {
    desc <- desc::description$new(base_path)
    as.list(desc$get(desc$fields()))$Package
  }
  else {
    fs::path_file(base_path)
  }
}

#' @rdname cranlogs_badge
#' @export
use_cranlogs_badge = function(
  summary = c("last-month", "last-day", 
              "last-week", "grand-total"),
  color = "blue") {
  summary <- match.arg(summary)
  if ( !(requireNamespace("fs", quietly = TRUE) &&
         requireNamespace("usethis", quietly = TRUE) &&
         requireNamespace("desc", quietly = TRUE) &&
         requireNamespace("rprojroot", quietly = TRUE)) ) {
    stop("Cannot use use_cranlogs_badge without fs, usethis, ", 
         "desc, and rprojroot packages")
  }
  package_name = project_name()
  
  pkg_url <- paste0("https://r-pkg.org/pkg/", package_name)
  
  badge_url <- paste0("https://cranlogs.r-pkg.org/badges/",
                      summary, "/",
                      package_name, 
                      "?color=", color)

  usethis::use_badge("CRAN RStudio mirror downloads", pkg_url, badge_url)
}


R/cranlogs.R

#' @importFrom httr GET content stop_for_status
#' @importFrom jsonlite fromJSON


base_url  <- "http://cranlogs.r-pkg.org/"
daily_url <- paste0(base_url, "downloads/daily/")
top_url   <- paste0(base_url, "top/")

#' Daily package downloads from the RStudio CRAN mirror
#'
#' @param packages A character vector, the packages to query,
#'   or \code{NULL} for a sum of downloads for all packages.
#'   Alternatively, it can also be \code{"R"}, to query downloads
#'   of R itself. \code{"R"} cannot be mixed with packages.
#' @param when \code{last-day}, \code{last-week} or \code{last-month} (see 
#' details). If this is given, then \code{from} and \code{to} are ignored.
#' @param from Start date, in \code{yyyy-mm-dd} format, or
#'   \code{last-day}. It is ignored if \code{when} is given.
#' @param to End date, in \code{yyyy-mm-dd} format, or
#'   \code{last-day}. It is ignored if \code{when} is given.
#' @param batch_size Maximum batch size per query.
#' A batch size <=800 is reccomended, as going above that can cause errors.
#' @param mc.cores Number of cores to parallelise batched queries across.
#' @return For packages a data frame with columns:
#'   \item{\code{package}}{The package. This column is missing if
#'     all packages were queried.}
#'   \item{\code{date}}{Day of the downloads, it is a Date object.}
#'   \item{\code{count}}{Download count.}
#'
#'   For downloads of R, there are also columns for the operating
#'   system (\code{os}) and the R version (\code{version}).
#'
#' @details \code{last-day} is the last day for which data is available,
#'  \code{last-week} is from 6 days prior to that last day with data, 
#'  \code{last-month} is from 29 days prior to that last day with data.
#'  
#'  0 counts can be due to the non-availability of data on the RStudio server 
#'  for that day.
#' @family CRAN downloads
#' @export
#' @importFrom parallel mclapply
#' @examples 
#' ## Default is last day for which data is available.
#' cran_downloads()
#'
#' ## All downloads for 'dplyr' in the last day for which data is available.
#' cran_downloads(packages = "dplyr")
#'
#' ## Daily downloads for 'igraph' last week
#' ## (6 days prior to the last day for which data is available)
#' cran_downloads(packages = "igraph", when = "last-week")
#'
#' ## Downloads in the specified time interval
#' cran_downloads(from = "2014-06-30", to = "2014-08-08")
#'
#' ## Multiple packages
#' cran_downloads(packages = c("ggplot2", "plyr", "dplyr"))
#'
#' ## R downloads
#' cran_downloads("R") 
cran_downloads <- function(packages = NULL,
                           when = c("last-day", "last-week", "last-month"),
                           from = "last-day", 
                           to = "last-day",
                           batch_size = 800,
                           mc.cores = 1) {

  if (!missing(when)) {
    interval <- match.arg(when)
  } else {
    when <- when[[1]] ### Take only first entry
    if (as.character(from) != "last-day") {
      check_date(from)
    }
    
    if (as.character(to) != "last-day") {
      check_date(to)
    }
    
    if (from == to) {
      interval <- from
    } else {
      interval <- paste(from, sep = ":", to)
    }
  } 
  #### Check packages ####
  if (!is.null(packages) &&
      "R" %in% packages && 
      any(packages != "R")) {
    stop("R downloads cannot be mixed with package downloads")
  }  
  #### Split into batches ####
  t1 <- Sys.time()
  batches <- split_batches(v = packages, 
                           batch_size = batch_size)
  df <- parallel::mclapply(seq_len(length(batches)), function(i){
    b <- batches[[i]]
    message_parallel(paste("Batch:",i,"/",length(batches)))
    query(interval=interval, 
          packages=b)
  }, mc.cores = mc.cores) 
  #### Bind all batch results ####
  df <- do.call(rbind,df)
  ### Report total time taken ####
  round(difftime(Sys.time(),t1),1)
  return(df)
}

query <- function(interval, 
                  packages){
  #### Prepare URI  ####
  if (is.null(packages)) {
    ppackages <- ""
  } else {
    ppackages <- paste(packages, collapse = ",")
    ppackages <- paste0("/", ppackages)
  } 
  #### Request data ####
  req <- GET(paste0(daily_url, interval, ppackages),
             httr::user_agent("cranlogs R package by R-hub"))
  #### Handle errors due to too many packages ###
  status <- tryCatch({
    stop_for_status(req)
  }, error=function(e){e})
  if(grepl("Request-URI Too Long (HTTP 414)",as.character(status))){
    stop("Request-URI Too Long (HTTP 414): Try reducing `batch_size`.")
  }
  #### Access content ####
  r <- fromJSON(content(req, as = "text"), simplifyVector = FALSE)
  #### Check content ####
  if ("error" %in% names(r) && r$error == "Invalid query") {
    stop("Invalid query, probably invalid dates")
  }
  #### Convert to data.frame ####
  to_df(r, packages)
}

to_df <- function(res, packages) {
  if (length(res) == 1 && identical(toupper(packages), "R")) {
    to_df_r(res[[1]])
  } else if (length(res) == 1 && is.null(res[[1]]$package)) {
    to_df_1(res[[1]])
  } else {
    dfs <- lapply(res, to_df_1)
    for (i in seq_along(res)) dfs[[i]]$package <- res[[i]]$package
    do.call(rbind, dfs)
  }
}

to_df_1 <- function(res1) {
  df <- data.frame(
    stringsAsFactors = FALSE,
    date = as.Date(vapply(res1$downloads, "[[", "", "day")),
    count = vapply(res1$downloads, "[[", 1, "downloads")
  )
  fill_in_dates(df, as.Date(res1$start), as.Date(res1$end))
}

to_df_r <- function(res1) {
  df <- data.frame(
    stringsAsFactors = FALSE,
    date = as.Date(vapply(res1$downloads, "[[", "", "day")),
    version = vapply(res1$downloads, "[[", "", "version"),
    os = vapply(res1$downloads, "[[", "", "os"),
    count = vapply(res1$downloads, "[[", 1, "downloads")
  )
  df
}

fill_in_dates <- function(df, start, end) {
  if (start > end) stop("Empty time interval")
  if (end > Sys.Date()) warning("Time interval in the future")

  dates <- seq(start, end, by = as.difftime(1, units = "days"))
  if (any(! dates %in% df$date)) {
    df2 <- data.frame(
      stringsAsFactors = FALSE,
      date = dates[! dates %in% df$date],
      count = 0
    )
    df <- rbind(df, df2)
    df <- df[order(df$date),]
    rownames(df) <- NULL
  }
  df
} 

#' Top downloaded packages from the RStudio CRAN mirror
#'
#' @param when \code{last-day}, \code{last-week} or \code{last-month} (see 
#' details).
#' @param count Number of packages to list. Note that the DB server
#'   lists only at most 100 packages. This number might change in the
#'   future.
#' @return A data frame with columns: \code{rank}, \code{package},
#'   \code{count}, \code{from}, \code{to}.
#'
#' @family CRAN downloads
#' @export
#' @examples 
#' ## Default is last day for which data is available.
#' cran_top_downloads()
#'
#' ## Last week (6 days prior to the last day for which data is available) 
#' ## instead
#' cran_top_downloads(when = "last-week") 
#' 
#' @details \code{last-day} is the last day for which data is available,
#'  \code{last-week} is from 6 days prior to that last day with data, 
#'  \code{last-month} is from 29 days prior to that last day with data.
#'  
#'  0 counts can be due to the non-availability of data on the RStudio server 
#'  for that day.

cran_top_downloads <- function(when = c("last-day", "last-week",
                                 "last-month"), count = 10) {

  when <- match.arg(when)
  req <- GET(paste0(top_url, when, '/', count),
             httr::user_agent("cranlogs R package by R-hub"))
  stop_for_status(req)
  r <- fromJSON(content(req, as = "text", encoding = "UTF-8"), 
    simplifyVector = FALSE)
  
  df <- data.frame(
    stringsAsFactors = FALSE,
    rank = seq_along(r$downloads),
    package = vapply(r$downloads, "[[", "", "package"),
    count = as.integer(vapply(r$downloads, "[[", "", "downloads")),
    from = as.Date(r$start),
    to = as.Date(r$end)
  )

  if (nrow(df) != count) {
    warning("Requested ", count, " packages, returned only ", nrow(df))
  }

  df
}
